### PR TITLE
Silver/cli

### LINKED ISSUE #22: Complete -a CLI Option
Complete the `-a load_address` CLI option.

### LINKED ISSUE #23: Complete -n CLI Option
Complete the `-n prog_number` CLI option.

### LINKED ISSUE #24: Complete -dump CLI Option
Complete the function that executes the `-dump nbr_cycle` CLI option.

### LINKED ISSUE #25: Complete -h CLI Option
Complete the function that executes the -h CLI option. This option prints the usage description
```
USAGE
./corewar [-dump nbr_cycle] [[-n prog_number] [-a load_address] prog_name] ...
DESCRIPTION
-dump nbr_cycle
dumps the memory after the nbr_cycle execution (if the round isn’t already over) with the following format: 32 bytes/line in hexadecimal (A0BCDEFE1DD3...)
-n prog_number
sets the next program’s number. By default, the first free number in the parameter order
-a load_address
sets the next program’s loading address. When no address is
specified, optimize the addresses so that the processes are as far away from each other as possible. The addresses are MEM_SIZE modulo
```

### LINKED ISSUE #26: Complete file parser

### PR DESCRIPTION
**Pull request by rswhipple**
### WHAT WAS DONE

- added flag_t struct to handle parsing options
- all options have been completed
- champion struct has been updated
- parsing flow has been refactored
- helper.c file added
- directory for example .S and .cor files added

### CODE REVIEW

File: src/vm_parse.c [hunk 1]
```diff
@@ -22,49 +22,67 @@ modulo
...
 
-    return num_champions;
+    return head;    // return head of champion linked list
 }
 
+void print_usage()
+{
+    // print usage instructions
+}
+
```
Comment (rwkliu) [inline src/vm_parse.c:85]: Add the printf line to print the usage instructions

File: include/op.h [hunk 2]
```diff
@@ -105,11 +101,12 @@ typedef struct header_s {
...
     header_t *champ_header;       // header
-    int id;                       // id of the champ
-    int num_instuctions;          // number of instructions
-    op_t *instructions;           // instruction array
-    int registers[16];            // address of registers
-    int pc;                       // program counter
+    int id;                       // id of champ
+    int address;                  // address of champ
+    int num_inst;                 // number of instructions
+    unsigned long long *inst;     // instruction array
```
Comment (rwkliu) [inline include/op.h:107]: The instruction array may not need to be unsigned long long. I'm experimenting with this data structure.

**Review by rwkliu — COMMENTED** [review]
Great progress. 

Even though checks passed, the compiler generated errors.
src/champion.c: In function ‘read_file’:
2024-02-05T03:48:50.8942190Z src/champion.c:56:29: warning: passing argument 1 of ‘getline’ from incompatible pointer type [-Wincompatible-pointer-types]
2024-02-05T03:48:50.8944651Z    56 |     while ((bytes = getline(buf, BUF_SIZE, fd)) != -1) {
2024-02-05T03:48:50.8945351Z       |                             ^~~
2024-02-05T03:48:50.8945896Z       |                             |
2024-02-05T03:48:50.8946605Z       |                             char *
2024-02-05T03:48:50.8947472Z In file included from src/../include/op.h:6,
2024-02-05T03:48:50.8948004Z                  from src/champion.c:1:
2024-02-05T03:48:50.8948835Z /usr/include/stdio.h:645:45: note: expected ‘char ** restrict’ but argument is of type ‘char *’
2024-02-05T03:48:50.8950834Z   645 | extern __ssize_t getline (char **__restrict __lineptr,
2024-02-05T03:48:50.8952036Z       |                           ~~~~~~~~~~~~~~~~~~^~~~~~~~~
2024-02-05T03:48:50.8952861Z In file included from src/champion.c:1:
2024-02-05T03:48:50.8954771Z src/../include/op.h:36:18: warning: passing argument 2 of ‘getline’ makes pointer from integer without a cast [-Wint-conversion]
2024-02-05T03:48:50.8956168Z    36 | #define BUF_SIZE 512UL
2024-02-05T03:48:50.8956739Z       |                  ^~~~~
2024-02-05T03:48:50.8957440Z       |                  |
2024-02-05T03:48:50.8958056Z       |                  long unsigned int
2024-02-05T03:48:50.8958954Z src/champion.c:56:34: note: in expansion of macro ‘BUF_SIZE’
2024-02-05T03:48:50.8959727Z    56 |     while ((bytes = getline(buf, BUF_SIZE, fd)) != -1) {
2024-02-05T03:48:50.8960262Z       |                                  ^~~~~~~~
2024-02-05T03:48:50.8960702Z In file included from src/../include/op.h:6,
2024-02-05T03:48:50.8961228Z                  from src/champion.c:1:
2024-02-05T03:48:50.8962200Z /usr/include/stdio.h:646:46: note: expected ‘size_t * restrict’ {aka ‘long unsigned int * restrict’} but argument is of type ‘long unsigned int’
2024-02-05T03:48:50.8963152Z   646 |                           size_t *__restrict __n,
2024-02-05T03:48:50.8963599Z       |                           ~~~~~~~~~~~~~~~~~~~^~~
2024-02-05T03:48:50.8964487Z src/champion.c:56:44: warning: passing argument 3 of ‘getline’ makes pointer from integer without a cast [-Wint-conversion]
2024-02-05T03:48:50.8965438Z    56 |     while ((bytes = getline(buf, BUF_SIZE, fd)) != -1) {
2024-02-05T03:48:50.8965918Z       |                                            ^~
2024-02-05T03:48:50.8966455Z       |                                            |
2024-02-05T03:48:50.8966985Z       |                                            int
2024-02-05T03:48:50.8967431Z In file included from src/../include/op.h:6,
2024-02-05T03:48:50.8967907Z                  from src/champion.c:1:
2024-02-05T03:48:50.8968735Z /usr/include/stdio.h:647:44: note: expected ‘FILE * restrict’ but argument is of type ‘int’
2024-02-05T03:48:50.8969403Z   647 |                           FILE *__restrict __stream) __wur;
2024-02-05T03:48:50.8969918Z       |                           ~~~~~~~~~~~~~~~~~^~~~~~~~

**Review by rwkliu — CHANGES_REQUESTED** [review]
https://github.com/rwkliu/Core-War/pull/27#pullrequestreview-1862018646
Great progress.
I think the warnings and errors should be fixed before merging the changes. 
You can request another review for this pull request when you're done.

**Comment by rswhipple** [comment]
@rwkliu I'm not getting any errors when I compile, so it will be difficult for me to fix. The read_file function is in process so I'd rather merge this branch as is and start working from a new branch on that one.

**Comment by rwkliu** [comment]
@rswhipple I get the errors on my machine and on the CI test. [https://github.com/rwkliu/Core-War/actions/runs/7775811036/job/21210254871?pr=27](url) Can you move the read_file WIP code to another branch to get rid of the errors and warnings? Then we can merge everything else on this branch.

**Pull request closed without merging**